pkg/provision/storage/perWorkspaceStorage.go

```diff
@@ -0,0 +1,231 @@
+	return needsStorage(workspace)
+}
+
+func syncPerWorkspacePVC(workspace *dw.DevWorkspace, clusterAPI sync.ClusterAPI) (*corev1.PersistentVolumeClaim, error) {
+	namespace := workspace.Namespace
+
+	namespacedConfig, err := nsconfig.ReadNamespacedConfig(namespace, clusterAPI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read namespace-specific configuration: %w", err)
+	}
+
+	// TODO: Should we adjust the storage size?
+	pvcSize := constants.PVCStorageSize
+	if namespacedConfig != nil && namespacedConfig.CommonPVCSize != "" {
+		pvcSize = namespacedConfig.CommonPVCSize
+	}
+
+	// We have to call clusterAPI.Client.Scheme() because the fakeClusterAPI.Scheme used in testing returns nil
+	pvc, err := getPerWorkspacePVCSpec(workspace, clusterAPI.Client.Scheme(), pvcSize)
+	if err != nil {
+		return nil, err
+	}
+
+	currObject, err := sync.SyncObjectWithCluster(pvc, clusterAPI)
+	switch t := err.(type) {
+	case nil:
+		break
+	case *sync.NotInSyncError:
+		return nil, &NotReadyError{
+			Message: "Updated per-workspace PVC on cluster",
+		}
+	case *sync.UnrecoverableSyncError:
+		return nil, &ProvisioningError{
+			Message: "Failed to sync per-workspace PVC to cluster",
+			Err:     t.Cause,
+		}
+	default:
+		return nil, err
+	}
+
+	currPVC, ok := currObject.(*corev1.PersistentVolumeClaim)
+	if !ok {
+		return nil, errors.New("tried to sync per-workspace PVC to cluster but did not get a PVC back")
+	}
+
+	return currPVC, nil
+
+}
+
+func getPerWorkspacePVCSpec(workspace *dw.DevWorkspace, scheme *runtime.Scheme, size string) (*corev1.PersistentVolumeClaim, error) {
+	// TODO: Just make these part of the function parameters instead?
+	namespace := workspace.Namespace
+	DevWorkspaceId := workspace.Status.DevWorkspaceId
+
+	pvcStorageQuantity, err := resource.ParseQuantity(size)
+	if err != nil {
+		return nil, err
+	}
+
+	pvcName := common.PerWorkspacePVCName(DevWorkspaceId)
+
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pvcName,
+			Namespace: namespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				corev1.ReadWriteOnce,
+			},
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					"storage": pvcStorageQuantity,
+				},
+			},
+			StorageClassName: config.Workspace.StorageClassName,
+		},
+	}
+
+	err = controllerutil.SetControllerReference(workspace, pvc, scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return pvc, nil
+
+}
+
+func (p *perWorkspaceStorageProvisioner) ProvisionStorage(podAdditions *v1alpha1.PodAdditions, workspace *dw.DevWorkspace, clusterAPI sync.ClusterAPI) error {
+	// Add ephemeral volumes
+	if err := addEphemeralVolumesFromWorkspace(workspace, podAdditions); err != nil {
+		return err
+	}
+
+	// If persistent storage is not needed, we're done
+	if !p.NeedsStorage(&workspace.Spec.Template) {
+		return nil
+	}
```

Comment:
If you decide to follow the suggestion for returning `false` for `NeedsStorage`, this should be replaced with `needStorage` from common.go:

```suggestion
	// If persistent storage is not needed, we're done
	if !needsStorage(&workspace.Spec.Template) {
		return nil
	}
```